### PR TITLE
feature: eclipse-tractusx/traceability-foss#639 rework handling of ex…

### PR DESCRIPTION
**Pull request by ds-lcapellino**
…pired policies

## Pre-review checks

Please ensure to do as many of the following checks as possible, before asking for committer review:

- [x] DEPENDENCIES are up-to-date. [Dash license tool](https://github.com/eclipse/dash-licenses). Committers can open IP issues for restricted libs.
- [x] [Copyright and license header](https://eclipse-tractusx.github.io/docs/release/trg-7/trg-7-02) are present on all affected files

### CODE REVIEW

File: irs-edc-client/src/main/java/org/eclipse/tractusx/irs/edc/client/exceptions/UsagePolicyExpiredException.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+ */
+@Getter
+public class UsagePolicyExpiredException extends EdcClientException {
+
+    private final transient Policy policy;
+    private final String businessPartnerNumber;
+
+    public UsagePolicyExpiredException(final String itemId, final Policy policy, final String businessPartnerNumber) {
+        super("Consumption of asset '" + itemId
+                + "' is not permitted as the required catalog offer policies do not comply with defined IRS policies.");
```
Comment (mkanal) [inline irs-edc-client/src/main/java/org/eclipse/tractusx/irs/edc/client/exceptions/UsagePolicyExpiredException.java:36]:
```suggestion
                + "' is not permitted as the required catalog offer policies do not comply with defined policies.");
```

Comment (mkanal) [reply irs-edc-client/src/main/java/org/eclipse/tractusx/irs/edc/client/exceptions/UsagePolicyExpiredException.java:36]: IRS should not be mentioned here, Policy Store is a generic component for handling policies. Nothing to do with IRS in general.

File: irs-edc-client/src/main/java/org/eclipse/tractusx/irs/edc/client/exceptions/UsagePolicyPermissionException.java [hunk 2]
```diff
@@ -27,15 +27,16 @@
...
  */
 @Getter
-public class UsagePolicyException extends EdcClientException {
+public class UsagePolicyPermissionException extends EdcClientException {
 
     private final transient Policy policy;
     private final String businessPartnerNumber;
 
-    public UsagePolicyException(final String itemId, final Policy policy, final String businessPartnerNumber) {
+    public UsagePolicyPermissionException(final String itemId, final Policy policy,
```
Comment (mkanal) [inline irs-edc-client/src/main/java/org/eclipse/tractusx/irs/edc/client/exceptions/UsagePolicyPermissionException.java:38]: Why Permission Expecption - The Policy does not comply. IMHO permission is not the right scope as policy itself does not comply which includes the mismatch of one or more permissions which are bundled inside a policy.

File: irs-edc-client/src/main/java/org/eclipse/tractusx/irs/edc/client/policy/PolicyCheckerService.java [hunk 3]
```diff
@@ -44,20 +44,34 @@ public class PolicyCheckerService {
...
     public boolean isValid(final Policy policy, final String bpn) {
-        return policy.getPermissions().stream().allMatch(permission -> isValid(permission, getValidStoredPolicies(bpn)));
+        final List<AcceptedPolicy> validStoredPolicies = getValidStoredPolicies(bpn);
+        return policy.getPermissions()
+                     .stream()
+                     .allMatch(permission -> hasValidConstraints(permission, validStoredPolicies));
     }
 
-    private boolean isValid(final Permission permission, final List<AcceptedPolicy> validStoredPolicies) {
+    private boolean hasValidConstraints(final Permission permission, final List<AcceptedPolicy> validStoredPolicies) {
```
Comment (mkanal) [inline irs-edc-client/src/main/java/org/eclipse/tractusx/irs/edc/client/policy/PolicyCheckerService.java:53]: Introduction of  a new term "constraint" - Is permission not the right term to use here? What are constraints in this relation? What is checked here specifically the permission term a constraint inside the permission or the policy itself?

**Review by mkanal — CHANGES_REQUESTED** [review]

File: irs-edc-client/src/main/java/org/eclipse/tractusx/irs/edc/client/policy/PolicyCheckerService.java (hunk 3 above)
Comment (ds-lcapellino) [reply irs-edc-client/src/main/java/org/eclipse/tractusx/irs/edc/client/policy/PolicyCheckerService.java:53]: This is not a new term. Permissions consists out of constraints. We have to check all constraints of a permission against the configured ones in order to determine a valid policy as a whole.

**Review by mkanal — APPROVED** [review]
LGFM

**Pull request merged**